### DISCUSSION TITLE
Multi-platform Docker images

### DISCUSSION THREAD

**Discussion by namedgraph** [oxigraph/oxigraph#531] [General]
Hi,

Would it be possible to get Docker images for different platforms? I'm running on M1 Mac and this is what I get:
```bash
docker run --rm -v $PWD/data:/data -p 7878:7878 ghcr.io/oxigraph/oxigraph --location /data serve --bind 0.0.0.0:7878

WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
```
I manage to run it like this:
```bash
docker run --rm -v $PWD/data:/data -p 7878:7878 --platform linux/amd64 ghcr.io/oxigraph/oxigraph --location /data serve --bind 0.0.0.0:7878
```
But it seems super slow. I'm attempting to load less than 2K quads like this:
```bash
curl http://localhost:7878/store -H 'Content-Type: application/n-quads' -T ./test.nq
```
and it takes like 5 minutes. I guess it's not normal? :)

**Comment by Tpt** [comment]
Yes, that would be amazing! I need to figure out a way to build the arm64 image from the x86-64 Github Action platform that is not awfully slow. Maybe have the "build" part done using cross-compilation inside a x64-64 container and then move the built artifact to a aarch64 image?

> I guess it's not normal? :)

If it is running on an emulator it's not surprising. I believe docker is using qemu and not something like rosetta2.

 If you want something running right now, there is a homebrew tap of Oxigraph: https://github.com/oxigraph/homebrew-oxigraph

**Reply by Tpt** [reply]
I have found a good tutorial on Docker cross-compilation: https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/

**Reply by Tpt** [reply]
The new 0.3.17 version now provides an arm64 docker image

**Comment by namedgraph** [comment]
I think `docker buildx` is the way these days: [How to Rapidly Build Multi-Architecture Images with Buildx](https://www.docker.com/blog/how-to-rapidly-build-multi-architecture-images-with-buildx/)
But I haven't tried myself yet.

**Reply by Tpt** [reply]
Yes but it recommands to run qemu when building for aarch64 and an Oxigraph build for qemu on GitHub actions takes something like 2hs... I would love to avoid that.

**Reply by namedgraph** [reply]
Isn't qemu only recommended for Docker Desktop?

**Reply by Tpt** [reply]
It's also used a lot in CIs: https://stackoverflow.com/questions/70312490/github-actions-runner-environment-doesnt-build-for-arm-images and Docker maintains a Github action to use it easily: https://github.com/docker/setup-qemu-action/

It works very well for cheap processes (but not for heavy builds like RocksDB...)

**Reply by namedgraph** [reply]
There's something called self-hosted GH runners that I haven't heard about before. But they seem to support multiple architectures: https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners#architectures

**Reply by Tpt** [reply]
Yes, it might be very handy. But we need somewhere running ARM64 to host it. AWS is an option or maybe https://www.scaleway.com/en/amp2-instances/

**Reply by adamreichold** [reply]
I think tapping into the large existing pool of amd64 runners and using multi-stage cross compilation is definitely preferable. This would also support even more exotic and/or small platforms.

**Comment by namedgraph** [comment]
BTW I have [enabled Rosetta for the whole Terminal](https://apple.stackexchange.com/a/428769).

**Reply by Tpt** [reply]
Sadly I am not sure it is enough: https://github.com/docker/roadmap/issues/384

I believe you can check if docker is running qemu by checking in the "system monitor" is there is a related process.
